### PR TITLE
Allow for accessing classes without fully qualified name

### PR DESCRIPTION
**Pull request by inxilpro**
This is a port of the wonderful [laravel-tinker-tools](https://github.com/spatie/laravel-tinker-tools). Disabled by default and enabled by a configuration flag.

tinker

### CODE REVIEW

File: src/TinkerServiceProvider.php [hunk 1]
```diff
@@ -26,6 +26,10 @@ public function register()
         });
 
         $this->commands(['command.tinker']);
+
+        $this->publishes([
```
Comment (GrahamCampbell) [inline src/TinkerServiceProvider.php]: Won't work on lumen, and should only be called in cli mode.

File: src/Console/TinkerCommand.php [hunk 2]
```diff
@@ -53,6 +54,14 @@ public function fire()
         $shell->addCommands($this->getCommands());
         $shell->setIncludes($this->argument('include'));
 
+        if (config('tinker.guess_fqcn', false)) {
```
Comment (GrahamCampbell) [inline src/Console/TinkerCommand.php]: The config global function is not available to be used.

File: src/TinkerServiceProvider.php (hunk 1 above)
Comment (inxilpro) [inline src/TinkerServiceProvider.php]: I wrapped this in a `if ($this->app->runningInConsole())` but I don't use Lumen so I'm not sure what needs to be done to ensure compatibility…

File: src/Console/TinkerCommand.php (hunk 2 above)
Comment (inxilpro) [inline src/Console/TinkerCommand.php]: Got it. Switched to `$this->getLaravel()->make('config')`.

File: src/TinkerServiceProvider.php [hunk 3]
```diff
@@ -27,9 +27,11 @@ public function register()
 
         $this->commands(['command.tinker']);
 
-        $this->publishes([
-            __DIR__.'/../config/tinker.php' => $this->app['path.config'].DIRECTORY_SEPARATOR.'tinker.php',
-        ]);
+        if ($this->app->runningInConsole()) {
```
Comment (GrahamCampbell) [inline src/TinkerServiceProvider.php]: Still doesn't fix lumen support. Please see my blog post for what I mean. https://blog.bugsnag.com/designing-laravel-packages/. Feel free to lift the code from https://github.com/GrahamCampbell/Laravel-Example/blob/master/src/CatServiceProvider.php#L12-L23.

Comment (GrahamCampbell) [inline src/TinkerServiceProvider.php]: I actually don't think config is even needed here. I'd say just force the first setting on, and the second setting off.

Comment (inxilpro) [inline src/TinkerServiceProvider.php]: I'm fine with that. I just wasn't sure if everyone would want the functionality.

**Comment by GrahamCampbell** [comment]
Please also delete the config file. :)

File: src/Console/TinkerCommand.php [hunk 4]
```diff
@@ -53,6 +54,11 @@ public function fire()
         $shell->addCommands($this->getCommands());
         $shell->setIncludes($this->argument('include'));
 
+        (new ClassAliasAutoloader(
+            $shell, 
+            $this->getLaravel()->basePath().'/vendor/composer/autoload_classmap.php'
```
Comment (inxilpro) [inline src/Console/TinkerCommand.php]: @GrahamCampbell I had been using `base_path()` but I assume that helper shouldn't be relied upon here. This seem like a Laravel/Lumen compatible way to find the class map?

Comment (GrahamCampbell) [inline src/Console/TinkerCommand.php]: I think it looks ok. I think laravel 4 had a vendor folder constant, but I think we got rid of it, so this is fine, so long as `basePath` exists on both laravel and lumen application classes. :)

File: src/ClassAliasAutoloader.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    /**
+     * @var  Shell
+     */
+    protected $shell;
+    
+    /**
+     * Constructor
+     *
+     * @param  Shell  $shell
```
Comment (GrahamCampbell) [inline src/ClassAliasAutoloader.php]: Please use the full name here. :)

File: src/ClassAliasAutoloader.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+class ClassAliasAutoloader
+{
+    /**
+     * @var  array
+     */
+    protected $classes = [];
+
+    /**
+     * @var  Shell
```
Comment (GrahamCampbell) [inline src/ClassAliasAutoloader.php]: Same, and only one space needed. :)

File: src/ClassAliasAutoloader.php [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+     */
+    protected $shell;
+    
+    /**
+     * Constructor
+     *
+     * @param  Shell  $shell
+     * @param  string  $classMapPath
+     */
```
Comment (GrahamCampbell) [inline src/ClassAliasAutoloader.php]: Missing `@return void`.

**Comment by GrahamCampbell** [comment]
Looking pretty good. Just some minor things remaining. :)

**Comment by GrahamCampbell** [comment]
Then, Taylor can review. :)

**Comment by inxilpro** [comment]
OK, I think we should be good.

File: src/ClassAliasAutoloader.php [hunk 8]
```diff
@@ -7,20 +7,21 @@
...
+     * @var \Psy\Shell
      */
     protected $shell;
     
     /**
      * Constructor
      *
-     * @param  Shell  $shell
-     * @param  string  $classMapPath
+     * @param \Psy\Shell $shell
```
Comment (GrahamCampbell) [inline src/ClassAliasAutoloader.php]: Please do use 2 spaces on param annotations tho.

Comment (inxilpro) [inline src/ClassAliasAutoloader.php]: Ah, OK :)

File: src/ClassAliasAutoloader.php [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    }
+    
+    /**
+     * Register the SPL autoloader
+     *
+     * @return void
+     */
+    public function registerAutoloader()
+    {
+        spl_autoload_register([$this, 'aliasClass']);
```
Comment (GrahamCampbell) [inline src/ClassAliasAutoloader.php]: Just noticed we have a bug here. This should be unregistered on class destruct.

**Pull request merged**